zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1927,15 +1927,23 @@ private void doAsyncCreate(final String path, final byte[] data, final CreateMod
       final long startT, final ZkAsyncCallbacks.CreateCallbackHandler cb,
       final String expectedSessionId) {
     retryUntilConnected(() -> {
-      getExpectedZookeeper(expectedSessionId)
-          .create(path, data, ZooDefs.Ids.OPEN_ACL_UNSAFE, mode, cb,
-              new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, false) {
-                @Override
-                protected void doRetry() {
-                  doAsyncCreate(path, data, mode, System.currentTimeMillis(), cb,
-                      expectedSessionId);
-                }
-              });
+      final ZooKeeper zk;
+      try {
+        zk = getExpectedZookeeper(expectedSessionId);
+      } catch (ZkSessionMismatchedException e) {
+        // Use BADARGUMENTS rc to represent session mismatch error (invalid expectedSession)
+        // and not retry.
+        cb.processResult(KeeperException.Code.BADARGUMENTS.intValue(), path,
+            new ZkAsyncCallMonitorContext(_monitor, startT, 0, false), null);
+        throw e;
+      }
+      zk.create(path, data, ZooDefs.Ids.OPEN_ACL_UNSAFE, mode, cb,
+          new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, false) {
+            @Override
+            protected void doRetry() {
+              doAsyncCreate(path, data, mode, System.currentTimeMillis(), cb, expectedSessionId);
```

Comment:
Why we need these given we have already cancel it in the doRetry()?

Reply:
It is to handle the first call before it enters doRetry() and asyncThread. For the single thread calling, it is OK as the exception is thrown to the upstream caller and the caller won't be able to call waitForSuccess().
If it is multi threading: thread-1 calls asyncCreate(), and thread-2 waitForSuccess(), thread-2 will be waiting forever.
So I'd like to process it as when the sessions mismatch, we should treat the operation a failure.

Reply:
It's a good catch!. But I think you should catch-all exceptions, not only session mismatch. There is no other place that we can cancel the callback.

Reply:
@jiajunwang I seriously thought about catching a generic exception, but after digging into the code more, `getExpectedZookeeper()` is almost impossible to throw an exception other than `ZkSessionMismatchedException`. That's the reason I think catching `ZkSessionMismatchedException` is good enough and I just try...catch `getExpectedZookeeper` other than the whole block below. It's cleaner to understand we just want to handle exception for `getExpectedZookeeper`, excluding `create()`. The KeeperException error from create() is handled in the async retry thread.
```
try {
  getExpectedZookeeper(expectedSessionId).create();
} catch (Exception e) {
}
```
Maybe you could say catching a generic one is safer.

Reply:
I know the logic is absolutely correct. Just like when we first implemented this retry logic and there is no getExpectedZookeeper() call so nothing will go wrong.

The point is that we may have other exceptions if the new code is added here. Then this issue might happen if any user is using a different thread to wait.

Back to my original question, what is the problem of a generic try-catch? To be clear, what I mean is,

try {
  retryUntilConnect(....)
} catch (Exception ex) {
  handle callback
}

IMO, this looks cleaner : )

Reply:
@jiajunwang OK got you. I don't think there is a function problem. I just could not find a good KeeperException code for handling callback. Previously I thought of a similar code to catch the whole private API, which might be a bit cleaner? - same thing, wrapping inside the API vs outside.
```
public void asyncCreate() {
  try {
    doAsyncCreate();
  } catch (RuntimeException e) {
   // I think we'll need a generic code here but I am not sure what error code is a good one
    cb.processResult(KeeperException.Code.BADARGUMENTS.intValue());
    throw e;
  }
```
I can't find a generic error code for the generic runtimeException: `KeeperException.Code.BADARGUMENTS` cannot represent a generic RuntimeException. Maybe a customized one like -255? 
The reason I put catch for getExpectedZookeeper is we may use BADARGUMENTS to represent session mismatch.

Let me know your opinions, including what error code number is preferred.

Reply:
In this case, I think UNKNOWN_RET_CODE is fine. Because you really don't have it, right?

Reply:
Refactored. @jiajunwang Could you take another look? Thanks.